### PR TITLE
Type and size-safe bugfixes

### PR DESCRIPTION
**Pull request by jzulauf-lunarg**
Two bugfixes discovered during 32-bit type/size safety assessment.

### CODE REVIEW

File: layers/buffer_validation.cpp [hunk 1]
```diff
@@ -930,7 +930,7 @@ bool CoreChecks::ValidateAndUpdateQFOScoreboard(const debug_report_data *report_
                        " duplicates existing barrier submitted in this batch from command buffer %s.",
                        "vkQueueSubmit()", BarrierRecord::BarrierName(), operation, BarrierRecord::HandleName(),
                        report_data->FormatHandle(barrier.handle).c_str(), barrier.srcQueueFamilyIndex, barrier.dstQueueFamilyIndex,
-                       report_data->FormatHandle(inserted.first->second).c_str());
+                       report_data->FormatHandle(inserted.first->second->commandBuffer).c_str());
```
Comment (jzulauf-lunarg) [inline layers/buffer_validation.cpp]: FormatHandle isn't current typesafe w.r.t. requiring *VULKAN* handles... so this slipped through.

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -2039,7 +2039,7 @@ bool CoreChecks::ValidateDeviceMaskToRenderPass(GLOBAL_CB_NODE *pCB, uint32_t de
     if ((deviceMask & pCB->active_render_pass_device_mask) != deviceMask) {
         skip |= log_msg(report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VUID_handle_type, VUID_handle, VUID,
                         "deviceMask(0x%" PRIx32 ") is not a subset of the render pass[%s] device mask(0x%" PRIx32 ").", deviceMask,
-                        report_data->FormatHandle(pCB->activeRenderPass).c_str(), pCB->active_render_pass_device_mask);
+                        report_data->FormatHandle(pCB->activeRenderPass->renderPass).c_str(), pCB->active_render_pass_device_mask);
```
Comment (jzulauf-lunarg) [inline layers/core_validation.cpp]: Also, here.

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -13118,32 +13118,32 @@ VkResult CoreChecks::CoreLayerCreateValidationCacheEXT(VkDevice device, const Vk
 
 void CoreChecks::CoreLayerDestroyValidationCacheEXT(VkDevice device, VkValidationCacheEXT validationCache,
                                                     const VkAllocationCallbacks *pAllocator) {
-    delete (ValidationCache *)validationCache;
+    delete CastFromHandle<VkValidationCacheEXT, ValidationCache *>(validationCache);
```
Comment (jzulauf-lunarg) [inline layers/core_validation.cpp]: Ensure cast is size-safe, as on 32bit the pointer and the handle aren't.

File: layers/core_validation.cpp [hunk 4]
```diff
@@ -13118,32 +13118,32 @@ VkResult CoreChecks::CoreLayerCreateValidationCacheEXT(VkDevice device, const Vk
...
-    ((ValidationCache *)validationCache)->Write(pDataSize, pData);
+    CastFromHandle<VkValidationCacheEXT, ValidationCache *>(validationCache)->Write(pDataSize, pData);
     return (pData && *pDataSize != inSize) ? VK_INCOMPLETE : VK_SUCCESS;
 }
 
 VkResult CoreChecks::CoreLayerMergeValidationCachesEXT(VkDevice device, VkValidationCacheEXT dstCache, uint32_t srcCacheCount,
                                                        const VkValidationCacheEXT *pSrcCaches) {
     bool skip = false;
-    auto dst = (ValidationCache *)dstCache;
-    auto src = (ValidationCache const *const *)pSrcCaches;
```
Comment (jzulauf-lunarg) [inline layers/core_validation.cpp]: On 32bit this cast means that the ValidationCache array would alternate between valid pointers and nullptr's

File: layers/core_validation.cpp [hunk 5]
```diff
@@ -13118,32 +13118,32 @@ VkResult CoreChecks::CoreLayerCreateValidationCacheEXT(VkDevice device, const Vk
...
+    auto dst = CastFromHandle<VkValidationCacheEXT, ValidationCache *>(dstCache);
     VkResult result = VK_SUCCESS;
     for (uint32_t i = 0; i < srcCacheCount; i++) {
-        if (src[i] == dst) {
+        auto src = CastFromHandle<const VkValidationCacheEXT, const ValidationCache *>(pSrcCaches[i]);
```
Comment (jzulauf-lunarg) [inline layers/core_validation.cpp]: Need to access the array of 64 bit values as 64 bit values, then cast.

File: layers/shader_validation.cpp [hunk 6]
```diff
@@ -2532,11 +2532,10 @@ bool CoreChecks::ValidateRayTracingPipelineNV(PIPELINE_STATE *pipeline) {
 uint32_t ValidationCache::MakeShaderHash(VkShaderModuleCreateInfo const *smci) { return XXH32(smci->pCode, smci->codeSize, 0); }
 
 static ValidationCache *GetValidationCacheInfo(VkShaderModuleCreateInfo const *pCreateInfo) {
-    while ((pCreateInfo = (VkShaderModuleCreateInfo const *)pCreateInfo->pNext) != nullptr) {
-        if (pCreateInfo->sType == VK_STRUCTURE_TYPE_SHADER_MODULE_VALIDATION_CACHE_CREATE_INFO_EXT)
-            return (ValidationCache *)((VkShaderModuleValidationCacheCreateInfoEXT const *)pCreateInfo)->validationCache;
+    const auto validation_cache_ci = lvl_find_in_chain<VkShaderModuleValidationCacheCreateInfoEXT>(pCreateInfo->pNext);
```
Comment (jzulauf-lunarg) [inline layers/shader_validation.cpp]: Use the lvl utility and the size safe cast.

File: layers/vk_layer_utils.h [hunk 7]
```diff
@@ -159,6 +159,34 @@ static uint64_t CastToUint64(HandleType handle) {
     return static_cast<uint64_t>(CastToUint<HandleType, Uint>(handle));
 }
 
+// Convenience functions to case between handles and the types the handles abstract, reflecting the Vulkan handle scheme, where
```
Comment (jzulauf-lunarg) [inline layers/vk_layer_utils.h]: A couple more variants on size-safe casts to make the above more convenient.

File: layers/core_validation.cpp (hunk 3 above)
Comment (tobine) [reply layers/core_validation.cpp]: Note that with [template argument deduction](https://en.cppreference.com/w/cpp/language/template_argument_deduction) I believe the first template argument can be excluded in all of these calls.

File: layers/core_validation.cpp [hunk 8]
```diff
@@ -13118,32 +13118,32 @@ VkResult CoreChecks::CoreLayerCreateValidationCacheEXT(VkDevice device, const Vk
...
                                                     const VkAllocationCallbacks *pAllocator) {
-    delete (ValidationCache *)validationCache;
+    delete CastFromHandle<VkValidationCacheEXT, ValidationCache *>(validationCache);
 }
 
 VkResult CoreChecks::CoreLayerGetValidationCacheDataEXT(VkDevice device, VkValidationCacheEXT validationCache, size_t *pDataSize,
                                                         void *pData) {
     size_t inSize = *pDataSize;
-    ((ValidationCache *)validationCache)->Write(pDataSize, pData);
+    CastFromHandle<VkValidationCacheEXT, ValidationCache *>(validationCache)->Write(pDataSize, pData);
```
Comment (tobine) [inline layers/core_validation.cpp]: And for this guy I believe you can completely forgo the template args.

Comment (jzulauf-lunarg) [reply layers/core_validation.cpp]: I can eliminate one of them if I reorder the template arguments s.t. the formal parameter type is the second...

**Comment by locke-lunarg** [comment]
LGTM

**Pull request merged**